Patch for issue #19439, ActiveJob::Logger should check that logger formatter responds to `current_tags`

robdimarco opened the pull request:
When checking to see if the ActiveJob tag should be added to the logger, we need to retrieve the current list of tags. If the log formatter does not respond to the method `current_tags`, then we should assume the tag has already been added.

This addresses issue #19439

Commit:
Check that logger formatter responds to `current_tags`

cristianbica commented:
I don't think we should handle in rails any partial implementations of `ActiveSupport::TaggedLogging`. 
IMO we might merge a patch where we check for `current_tags` also but to condition the whole tagging behaviour. The example patch below cleans-up a little on the logger/formatter thing and checks for the `current_tags` method. This won't crash with partial tagging implementations. After (and if) this gets merged you can ask (or send a PR) to the syslogger maintainers to implement the `current_tags` method.

```diff
diff --git a/activejob/lib/active_job/logging.rb b/activejob/lib/active_job/logging.rb
index cd29e69..2687457 100644
--- a/activejob/lib/active_job/logging.rb
+++ b/activejob/lib/active_job/logging.rb
@@ -38,7 +38,7 @@ module ActiveJob
 
     private
       def tag_logger(*tags)
-        if logger.respond_to?(:tagged)
+        if logger.respond_to?(:tagged) && logger.respond_to?(:current_tags)
           tags.unshift "ActiveJob" unless logger_tagged_by_active_job?
           ActiveJob::Base.logger.tagged(*tags){ yield }
         else
@@ -47,7 +47,7 @@ module ActiveJob
       end
 
       def logger_tagged_by_active_job?
-        logger.formatter.current_tags.include?("ActiveJob")
+        logger.current_tags.include?("ActiveJob")
       end
 
     class LogSubscriber < ActiveSupport::LogSubscriber #:nodoc:
diff --git a/activesupport/lib/active_support/tagged_logging.rb b/activesupport/lib/active_support/tagged_logging.rb
index bcd7bf7..e685876 100644
--- a/activesupport/lib/active_support/tagged_logging.rb
+++ b/activesupport/lib/active_support/tagged_logging.rb
@@ -64,7 +64,7 @@ module ActiveSupport
       logger.extend(self)
     end
 
-    delegate :push_tags, :pop_tags, :clear_tags!, to: :formatter
+    delegate :push_tags, :pop_tags, :clear_tags!, :current_tags, to: :formatter
 
     def tagged(*tags)
       formatter.tagged(*tags) { yield self }
```

ziggythehamster commented:
If Rails is to allow custom loggers, I think it's wrong for Rails to assume they conform to any particular API unless Rails implements an abstract logger it expects every logger to inherit from. Thus, Rails should be checking that `logger` responds to any non-standard-library method. That said, Syslogger isn't entirely faultless, because they should have tested it against Rails 4.2 :).

I'm affected by this as well. What can I do to help get this through?

jetheredge commented:
We are being bit by this since we need to use Syslogger with Lograge. Unfortunately Syslogger exposes a `tagged

jetheredge commented:
I would also like to see if there is anything I can do to help get this change pushed through. We are unfortunately in a situation where we are using Syslogger and cannot wrap it in ActiveSupport::TaggedLogging. This keeps us from using ActiveJob due to the raised exception. ActiveJob appears to be the only place in Rails where logger.formatter.current_tags is ever checked. I agree with @cristianbica that pushing the method up to the logger level will allow the interface to be implemented more easily in Syslogger, but it seems to me that if the only method ActiveJob is using is `current_tags` then that method should be checked for as well.

arthurnn commented:
>This keeps us from using ActiveJob due to the raised exception. 

@jetheredge what you can do for now, is add a patch in our application to change Syslogger to respond to that `current_tags` method, of course thats a band-aid for now. But at least you wont be blocked.

Regarding the fix. I think the solution that @cristianbica proposed is pretty good. We should try to make the ActiveJob logger work with and without the Tagged feature.

jetheredge commented:
@arthurnn Thank you for the response. Yes, we have actually patched the logger_tagged_by_active_job? method to bypass the issue temporarily.

arthurnn commented:
@cristianbica do you want to give it a try with what you are proposing.
AND/OR @robdimarco do you wanna change this PR?

robdimarco committed:
Implement changes suggested by @cristianbica

robdimarco committed:
Add a delegation of :current_tags to the formatter

Per suggestion from @cristianbica add a delegation to the formatter

robdimarco committed:
Remove check within logger_tagged_by_active_job?

Implement suggestion from @cristianbica to check for non-standard methods in tag_logger not in logger_tagged_by_active_job?

robdimarco commented:
@arthurnn I made the changes suggested by @cristianbica

ffmike commented:
Personally, I'd like to see this merged. The combo of ActiveJob + Syslogger + Lograge is still busted due to ActiveJob assuming that anything implementing tagged also implements current_tags.

Meanwhile, if you're using Syslogger without tagging, you can work around the issue:

Syslogger.class_eval{|s|s.remove_possible_method :tagged}
